docs/errors.md

```diff
@@ -0,0 +1,249 @@
+
+## `MongoError`
+
+The base class from which all errors in the Node driver subclass.
+`MongoError` should **never** be be directly instantiated.
+![(MongoError hierarchy tree)](charts/imgs/MongoError.svg)
+
+Children of `MongoError` include:
+
+- [`MongoDriverError`](#MongoDriverError)
+- [`MongoNetworkError`](#MongoNetworkError)
+- [`MongoServerError`](#MongoServerError)
+- [`MongoSystemError`](#MongoSystemError)
+
+## `MongoDriverError`
+
+This class represents errors which originate in the driver itself or in the user's use of the driver.
+This class should **never** be directly instantiated.
+Its children are the main classes of errors that most users will interact with: [**`MongoLogicError`**](#MongoLogicError) and [**`MongoRuntimeError`**](#MongoRuntimeError).
+
+### `MongoLogicError`
+
+This class represents errors which originate from misuse of the driver API and will generally be thrown before making contact with the server.
+This class should **never** be directly instantiated.
+![(MongoLogicError hierarchy tree)](charts/imgs/MongoLogicError.svg)
+Children of `MongoLogicError` include:
+
+- #### `MongoInvalidArgumentError`
+
+  - Thrown when the user supplies malformed, unexpected arguments or failed to provide a required argument or field.
+
+- #### `MongoCompatibilityError`
+
+  - Thrown when a feature that is not enabled or allowed for the current configuration is used.
+
+- #### `MongoClientInstantiationError`
+
+  - Thrown when a user supplies a poorly-formatted connection string to the MongoClient constructor.
+
+- #### `MongoMissingCredentialsError`
+
+  - Thrown when a user fails to provide authentication credentials before attempting to connect to the mongo server.
+
+- #### `MongoDependencyError`
+  - Thrown when a required module or dependency is not present.
+
+### `MongoRuntimeError`
+
+This class represents errors which occur when the driver encounters unexpected input or reaches an unexpected/invalid internal state.
+This class should **never** be directly instantiated.
+
+_MongoRuntimeError children (pt 1)_
+![(MongoRuntimeError hierarchy tree part 1)](charts/imgs/MongoRuntimeError_0.svg)
+
+_MongoRuntimeError children (pt 2)_
+![(MongoRuntimeError hierarchy tree part 2)](charts/imgs/MongoRuntimeError_1.svg)
+
+_MongoRuntimeError children (pt 3)_
+![(MongoRuntimeError hierarchy tree part 3)](charts/imgs/MongoRuntimeError_2.svg)
+
+Children of `MongoRuntimeError` include:
+
+- #### `MongoInternalDriverError`
+
+  - Thrown when a catastrophic failure occurs in the driver.
+
+- #### `MongoTransactionError`
+
+  - Thrown when the user makes a mistake in the usage of transactions (e.g.: attempting to commit a transaction with a readPreference other than primary).
+
+- #### `MongoClientNotConnectedError`
+
+  - Thrown when the user attempts to operate on the data from a client that has not been connected to a MongoDB server instance.
+
+- #### `MongoKerberosClientConnectionError`
```

Comment:
`MongoKerberosClientConnectionError` -> `MongoKerberosConnectionError` / `MongoKerberosError`
I'm thinking that scoping this error for kerberos generally will serve us even if there's only the connection failure issue. But leaving connection in the name is also fine. I think we can drop Client since there's never a kerberos Server error on our end.